Improve batching work pool performance

bording opened the pull request:
This is the first PR mentioned in #251.

Currently this is based on my vs-support branch from #248, once that's merged, I'll rebase this against master instead.

https://github.com/rabbitmq/rabbitmq-dotnet-client/commit/f36ca331c217cfbc06aaf113ccdc242755bd4932 is what is new in this PR.

michaelklishin commented on projects/client/RabbitMQ.Client/src/client/impl/ConsumerWorkService.cs:
```diff
@@ -68,11 +62,21 @@ public void AddWork(IModel model, Action fn)
 
         public void RegisterKey(IModel model)
         {
+            if (model.ChannelNumber == 0)
```
For those confused by this magic constant: channel 0 is used for "special purpose" protocol communication.

bording replied:
Yes, I noticed that the current design causes even the `MainSession` model to be registered with the `ConsumerWorkService`, even though no consumers can be registered on it. Seemed like a reasonable optimization to skip those registrations.

michaelklishin replied:
Makes sense. Please leave a comment explaining that, through.

bording commented:
@michaelklishin I pushed up a revised commit with a comment about model 0, let me know if it's ok.

bording commented:
@kjnilsson What race condition was https://github.com/rabbitmq/rabbitmq-dotnet-client/commit/166cf17f0c981cdf244392a7efd687b659a24b41 trying to fix? I was testing this PR initially against 3.6.5, which didn't have this, and it looks like the ChannelNumber 0 optimization I made here is now throwing an NRE because of it.

Same thing problem should be occuring on #253.

bording commented:
I've rebased this branch on top of the lastest vs-support branch changes, and I went ahead and removed the special model 0 handling that was throwing an exception because of the ordering change. Having that in this branch is much less critical.

kjnilsson commented:
The concourse build failed on this test:

```
Errors and Failures
1) Failed : RabbitMQ.Client.Unit.TestConsumerOperationDispatch.TestDeliveryOrderingWithSingleChannel
  Expected: 96
  But was:  100
at RabbitMQ.Client.Unit.TestConsumerOperationDispatch.TestDeliveryOrderingWithSingleChannel() in /tmp/build/b95f7b06/rabbitmq_dotnet_client/projects/client/Unit/src/unit/TestConsumerOperationDispatch.cs:line 142
```

michaelklishin commented:
@kjnilsson that's likely a timing issue.

kjnilsson commented:
@michaelklishin yes just wanted to note it as it isn't yet public. Running the tests again.

bording commented:
This branch is now rebased against master.

For the failing test, yes that is expected on this PR, because the changes to the BatchingWorkPool can't guarantee ordering. Looking at the test and thinking about it more, I can see how this likely isn't an acceptable trade-off, given how acking of multiple delivery tags works.

kjnilsson commented:
@bording can you explain where the ordering guarantees are given up with your approach?

michaelklishin commented:
If we really give up the ordering guarantee in this PR, it's a no go compared to #253.

bording commented:
@kjnilsson Take a look at `BatchingWorkPool.AddWorkItem` and how it uses the `ready` queue. It will add an entry in there for every call, so you can end up with multiple `NextWorkBlock` calls looking at the same model's work.

The existing implementation uses the `SetQueue` class to prevent the same model from being queued multiple times, but the class isn't thread-safe, and I couldn't find any sort of concurrent version of a collection that had the same behavior.

@scooletz and I tried to come up with another way to prevent the same model from being queued multiple times, but everything we came up with either required reintroducing locks (and defeating the point of this change) or could potentially run into the opposite problem of having work items and not scheduling a `NextWorkBlock` call to dispatch them.

michaelklishin commented:
@bording then it sounds like you and @scooletz consider #253 objectively superior to this PR?

bording commented:
@michaelklishin If ordering is a hard requirement that can't be relaxed, then I think #253 is the only option, unless there's some way we haven't thought of to bring it back to this PR without reintroducing a lock.

 I only realized today that it might not be possible to give up ordering because of the multiple delivery tag ack option. We don't use that in our code at all, and I had forgotten that it was even part of the spec. :cry: 

If ordering wasn't a requirement, than I actually prefer this approach, because it does maintain the ability to pass in a custom task scheduler for concurrency throttling, and since it's not using dedicated threads that block when they don't have work, you see a lot less "synchronization time" when profiling compared to #253.

Given that this approach does have some benefits, that was why we initially presented both options.

michaelklishin commented:
We cannot give up ordering for backwards compatibility, too. I think the numbers you posted from #253 are very comparable. Not sure how many users provide a custom task scheduler.

kjnilsson commented:
Ok I've had a play with the PR to see if I can get the `TestDeliveryOrderingWithSingleChannel` to pass somehow but even after adding back the SetQueue (or SeqQueue behaviour rather) _and_ adding a bunch of relatively course grained locks the test still fails. I suspect something else may be going on that isn't immediately obvious. I now agree we should proceed with https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/253 . It does add a breaking change to a public interface (hello v5!) and a fair bit of internal changes as well but the approach is simple and easy to reason about. I don't think that many use a custom TaskScheduler but with a one to one relationship between `Model`s and threads it may not be necessary. 
Also it looks like it may allocate less than a creating tasks recursively which would be nice. Once 253 is rebased I will review it properly and test it.

michaelklishin commented:
Speaking of versions, we should probably release `4.1.0` with it (the API change is not going to affect too many).

michaelklishin commented:
Closing this in favour of #253 based on a number of discussions in #251, #253 and here.
